src/Xrefcheck/Scan.hs

```diff
@@ -61,12 +63,15 @@ gatherRepoInfo rw formatsSupport config root = do
   _ Tree.:/ repoTree <- liftIO $ Tree.readDirectoryWithL processFile rootNE
   let fileInfos = filter (\(path, _) -> not $ isIgnored path)
         $ dropSndMaybes . F.toList
-        $ Tree.zipPaths . (dirOfRoot Tree.:/)
+        $ Tree.zipPaths . (dirOfRoot root Tree.:/)
         $ filterExcludedDirs root repoTree
   return $ RepoInfo (M.fromList fileInfos)
   where
     rootNE = if null root then "." else root
-    dirOfRoot = if root == "" || root == "." then "" else takeDirectory root
+    dirOfRoot rt
+      | rt == "" || rt == "." = ""
+      | Unsafe.last rt == '/' = dirOfRoot $ takeDirectory rt
```

Comment:
I would use `Data.List.uncons` - it is simpler/quicker than `takeDirectory`, and safer than `Unsafe.last` (indeed using `Unsafe.last` is safe here, but it is always better to rely on compiler in ensuring correctness, this makes future edits less fragile).

Moreover, correct me if I'm wrong, but looks like a bug lurks here: according to the documentation, `takeDirectory "/"` is `"/"`, which means that passing `"/"` as a root would cause an infinite recursion.

Reply:
Oh, you're right, this definition of `dirOfRoot` does indeed produce an infinite loop on `"/"`! I'll read the rest of the comments and think about how to redefine the function.